packages/context/src/context.ts

```diff
@@ -34,16 +52,142 @@ export class Context {
+   */
+  private pendingEvents = 0;
+
+  /**
+   * Create a new context. For example,
+   * ```ts
+   * // Create a new root context, let the framework to create a unique name
+   * const rootCtx = new Context();
+   *
+   * // Create a new child context inheriting bindings from `rootCtx`
+   * const childCtx = new Context(rootCtx);
+   *
+   * // Create another root context called "application"
+   * const appCtx = new Context('application');
+   *
+   * // Create a new child context called "request" and inheriting bindings
+   * // from `appCtx`
+   * const reqCtx = new Context(appCtx, 'request');
+   * ```
    * @param _parent The optional parent context
+   * @param name Name of the context, if not provided, a `uuid` will be
+   * generated as the name
    */
   constructor(_parent?: Context | string, name?: string) {
+    super();
     if (typeof _parent === 'string') {
       name = _parent;
       _parent = undefined;
     }
     this._parent = _parent;
     this.name = name || uuidv1();
+
+    this.setupEventHandlers();
+  }
+
+  /**
+   * Wrap the debug statement so that it always print out the context name
+   * as the prefix
+   * @param formatter
+   * @param args
+   */
+  // tslint:disable-next-line:no-any
+  private _debug(formatter: any, ...args: any[]) {
+    /* istanbul ignore if */
+    if (debug.enabled) {
+      if (typeof formatter === 'string') {
+        debug(`[%s] ${formatter}`, this.name, ...args);
+      } else {
+        debug('[%s] ', this.name, formatter, ...args);
+      }
+    }
+  }
+
+  /**
+   * Set up an internal listener to notify registered observers asynchronously
+   * upon `bind` and `unbind` events
+   */
+  private setupEventHandlers() {
+    // The following are two async functions. Returned promises are ignored as
+    // they are long-running background tasks.
+    this.startNotificationTask('bind').catch(err => {
+      // Catch error to avoid lint violations
+      debug(err);
+      this.emit('error', err);
+    });
+    this.startNotificationTask('unbind').catch(err => {
+      // Catch error to avoid lint violations
+      debug(err);
+      this.emit('error', err);
+    });
+  }
+
+  /**
+   * Start a background task to listen on context events and notify observers
+   * @param eventType Context event type
+   */
+  private async startNotificationTask(eventType: ContextEventType) {
+    const notificationEvent = `${eventType}-notification`;
+    this.on(eventType, binding => {
+      // No need to schedule notifications if no observers are present
+      if (this.observers.size === 0) return;
+      // Track pending events
+      this.pendingEvents++;
+      // Take a snapshot of current observers to ensure notifications of this
+      // event will only be sent to current ones. Emit a new event to notify
+      // current context observers.
+      this.emit(notificationEvent, {
+        binding,
+        observers: new Set(this.observers),
```

Comment:
In your implementation, you are trading off performance of event notification for performance of adding a new observer:
 - when a new observer is added, we preserve the Set instance in `this.observers`
 - when a new event is fired, a new Set instance is created

I expect that events will be triggered more often than new observers are added, therefore we should do a different trade-off:
 - when a new observer is added, a new Set instance is created, it replaces the value stored in `this.observers`
 - when a new event is fired, we take a reference to the Set instance in `this.observers` because this instance is considered as immutable

Feel free to fix this in a follow-up pull request if you prefer.

Reply:
A new notification event is only fired when there are observers. I'll defer this micro-optimization.